src/main/scala/io/iohk/ethereum/db/storage/ReferenceCountNodeStorage.scala

```diff
@@ -36,99 +36,106 @@ class ReferenceCountNodeStorage(nodeStorage: NodeStorage, pruningOffset: BigInt,
-        .orElse(nodeStorage.get(nodeKey).map(storedNodeFromBytes)) // or db
-        .map(_.decrementReferences(1))
-
-      if (storedNode.isDefined) {
-        val pruneCanditatesUpdated =
-          // if references is 0, mark it as prune candidate for current block tag
-          if (storedNode.get.references == 0) toDeleteInBlock.copy(nodeKeys = nodeKey +: toDeleteInBlock.nodeKeys)
-          else toDeleteInBlock
-        (storedNodes + (nodeKey -> storedNode.get), pruneCanditatesUpdated)
+    val changes = toRemove.foldLeft(upsertChanges) { (storedNodes, nodeKey) =>
+      val maybeStoredNode: Option[(StoredNode, StoredNodeSnapshot)] = storedNodes.get(nodeKey) // get from current changes
+        .orElse(nodeStorage.get(nodeKey).map(storedNodeFromBytes).map(s => s -> StoredNodeSnapshot(nodeKey, Some(s)))) // or db
+
+      if (maybeStoredNode.isDefined) {
+        val (storedNode, snapshot) = maybeStoredNode.get
+        storedNodes + (nodeKey -> (storedNode.decrementReferences(1), snapshot))
       }
-      else acc
+      else storedNodes
     }
 
-    // map stored nodes to bytes in order to save them
-    val toUpsertUpdated = changes._1.map {
-      case (nodeKey: NodeHash, storedNode: StoredNode) => nodeKey -> storedNodeToBytes(storedNode)
-    }.toSeq
+    val (toRemoveUpdated, toUpsertUpdated, snapshots) =
+      changes.foldLeft(Seq.empty[NodeHash], Seq.empty[(NodeHash, NodeEncoded)], Seq.empty[StoredNodeSnapshot]) {
+        case ((removeAcc, upsertAcc, snapshotAcc), (key, (storedNode, theSnapshot))) =>
+          // If no more references, move it to the list to be removed
+          if(storedNode.references == 0) (removeAcc :+ key, upsertAcc, snapshotAcc :+ theSnapshot)
+          else (removeAcc, upsertAcc :+ (key -> storedNodeToBytes(storedNode)), snapshotAcc :+ theSnapshot)
+      }
 
-    val toMarkAsDeleted =
-      // Update prune candidates for current block tag if it references at least one block that should be removed
-      if (changes._2.nodeKeys.nonEmpty) Seq(toPruneInThisBlockKey -> pruneCandiatesToBytes(changes._2))
+    val snapshotToSave = {
+      if(snapshots.nonEmpty) {
+        // If not empty, snapshots will be stored indexed by block number and index
+        val snapshotCountKey = getSnapshotsCountKey(bn)
+        val blockNumberSnapshotsCount: BigInt = nodeStorage.get(snapshotCountKey).map(snapshotsCountFromBytes).getOrElse(0)
+        val snapshotsToSave = snapshots.zipWithIndex.map { case (snapshot, index) =>
+          getSnapshotKey(bn, blockNumberSnapshotsCount + index + 1) -> snapshotToBytes(snapshot)
+        }
+        // Save snapshots and latest snapshot index
+        (snapshotCountKey -> snapshotsCountToBytes(blockNumberSnapshotsCount + snapshotsToSave.size)) +: snapshotsToSave
+      }
       else Nil
-
-    nodeStorage.update(Nil, toUpsertUpdated ++ toMarkAsDeleted)
-
+    }
+    nodeStorage.update(toRemoveUpdated, toUpsertUpdated ++ snapshotToSave)
```

Comment:
As we are deleting some nodes here, shouldn't we somehow decrement the reference count of nodes this deleted node was referencing?  ( suppose we have node A referencing node B,  shouldn't we decrement reference count of B after deleting A)

Reply:
That a good question. In this context, references means that someone has "inserted" the node, which means, it's being used somewhere else. 

Think of a tree where nodes are chained by hash, so we might have, for example, two leafs with same value and those will be stored just once. When removing one of those leafs because we deleted the whole branch, we cannot physically remove it as there is another leaf equal to that one on a non deleted branch.

Reply:
And, specifically talking about your question, the storage only handles data being stored, not the trie logic.

So, if you have dependent nodes in a tree, the one using this storage (`io.iohk.ethereum.mpt.MerklePatriciaTrie`) will make sure references are consistent. Without getting too deep into the MPT, if you remove `A` and not `B` it will happen that:
- `A` references is decremented by 1
- `B` references is decremented by 1 
- `B` is linked to another node, let's call it `C` (which is either new or an existing one). It's is basically incrementing `B` references by 1
- `C` references is incremented by 1

So, `B` references are not changing in your scenario